cmd/lk/cloud.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (a *AuthClient) GetVerificationToken(deviceName string) (*VerificationToken, error) {
+	reqURL, err := url.Parse(a.baseURL + createTokenEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	params := url.Values{}
+	params.Add("device_name", deviceName)
+	reqURL.RawQuery = params.Encode()
+
+	resp, err := a.client.Post(reqURL.String(), "application/json", nil)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New(resp.Status)
+	}
+
+	err = json.NewDecoder(resp.Body).Decode(&a.verificationToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return &a.verificationToken, nil
+}
+
+func (a *AuthClient) ClaimCliKey(ctx context.Context) (*config.AccessKey, error) {
+	if a.verificationToken.Token == "" || time.Now().Unix() > a.verificationToken.Expires {
+		return nil, errors.New("session expired")
+	}
+
+	reqURL, err := url.Parse(a.baseURL + claimSessionEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	params := url.Values{}
+	params.Add("t", a.verificationToken.Token)
+	reqURL.RawQuery = params.Encode()
+
+	req, err := http.NewRequestWithContext(ctx, "POST", reqURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	resp, err := a.client.Do(req)
+	if resp != nil && resp.StatusCode == 404 {
+		return nil, errors.New("access denied")
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode == http.StatusUnauthorized {
+		// Not yet approved
+		return nil, nil
+	}
+
+	ak := &config.AccessKey{}
+	err = json.NewDecoder(resp.Body).Decode(&ak)
+	if err != nil {
+		return nil, err
+	}
+
+	return ak, nil
+}
+
+func (a *AuthClient) Deauthenticate() error {
+	// TODO: revoke any session token
+	return nil
+}
+
+func NewAuthClient(client *http.Client, baseURL string) *AuthClient {
+	a := &AuthClient{
+		client:  client,
+		baseURL: baseURL,
+	}
+	return a
+}
+
+func initAuth(ctx context.Context, cmd *cli.Command) error {
+	if err := loadProjectConfig(ctx, cmd); err != nil {
+		return err
+	}
+	authClient = *NewAuthClient(&http.Client{}, serverURL)
+	return nil
+}
+
+func handleAuth(ctx context.Context, cmd *cli.Command) error {
+	if disconnect {
+		return authClient.Deauthenticate()
+	}
+	return tryAuthIfNeeded(ctx, cmd)
+}
+
+func tryAuthIfNeeded(ctx context.Context, cmd *cli.Command) error {
+	_, err := loadProjectDetails(cmd)
+	if err != nil {
+		return err
+	}
+
+	// name
+	var deviceName string
+	if err := huh.NewInput().
+		Title("What is the name of this device?").
+		Value(&deviceName).
+		WithTheme(theme).
+		Run(); err != nil {
+		return err
+	}
+	fmt.Println("Device:", deviceName)
+
+	// request token
+	fmt.Println("Requesting verification token...")
+	token, err := authClient.GetVerificationToken(deviceName)
+	if err != nil {
+		return err
+	}
+
+	authURL, err := generateConfirmURL(token.Token)
+	if err != nil {
+		return err
+	}
+
+	// poll for keys
+	fmt.Printf("Please confirm access by visiting:\n\n   %s\n\n", authURL.String())
```

Comment:
it's also possible to launch this page too: [browser.OpenURL](https://github.com/livekit/livekit-cli/blob/main/cmd/lk/egress.go#L756)

Reply:
🆒 I didn't see that. Is it well-supported? Guessing it uses `xdg-open` on linux, `open` on mac, `start` on windows.

Reply:
Probably appropriate to use both methods, in case user has some popup blocking settings or smth.

Reply:
yah, always good to show the URL, but this has worked pretty well thus far.